R/dependencies.R

```diff
@@ -0,0 +1,410 @@
+      if (!db_chd$is_latest[i]) { ## ..not latest..
+        latest_id <- report_id ## ..use the pinned id
+      } else { ## pinned to latest
+        latest_id <- report_id
+      }
+    }
+
+
+    ## we return a dataframe with columns:
+    ## filename - the filename of the artefact
+    ## file_hash - the hash of the artefact
+    par_qry <- sprintf("report_version.id='%s'", latest_id)
+    sql_par_qry <-
+      paste(c("SELECT", "file_artefact.filename,",
+                        "file_artefact.file_hash",
+             "FROM", "(report_version",
+             "INNER JOIN", "report_version_artefact", "ON",
+                           "report_version.id = report_version_artefact.report_version",
+             "INNER JOIN", "file_artefact", "ON",
+                           "file_artefact.artefact = report_version_artefact.id)",
+             "WHERE", par_qry), collapse = " ")
+
+    db_par <- DBI::dbGetQuery(con, sql_par_qry)
+    i <- which(db_par$filename == filename)
+    if (db_par$file_hash[i] != file_hash)
+      return(TRUE)
+  }
+  return(FALSE)
+}
+
+##' @title Recursively check that none of the parents share the same name as the
+##' current report.
+##'
+##' @param parent_vertex The vertex whose name we want to check
+##' @param name The name we want to match
+##'
+##' @return A boolean TRUE if there is a match (bad!) FALSE if the is no match
+##' (good!)
+check_parents <- function(parent_vertex, name) {
+  if (parent_vertex$name == name) {
+    return(TRUE)
+  } else {
+    if (!is.null(parent_vertex$parent)) {
+      return(check_parents(parent_vertex$parent, name))
+    }
+  }
+  return(FALSE)
+}
+
+##' @title Recursively builds a tree for a given report
+##'
+##' @param name the name of the report
+##' @param id the id of the report, if omitted, use the id of the latest report
+##' @param depth [internal] - only used ensure we don't get trapped in an
+##'              infinite loop
+##' @param parent [internal] - the previous vertex in the tree
+##' @param tree [internal] - The tree object that is built up and returned at
+##'             the end
+##' @param con A connection to a database
+##' @param root
+##' @param locate
+##' @param upstream A boolean indicating if we want to move up or down the tree
+##'
+##' @return An R6 tree object
+build_tree <- function(name, id, depth = 100, parent = NULL,
+                       tree = NULL, con, upstream = FALSE, list_all = FALSE) {
+  ## this should never get triggered - it only exists the prevent an infinite
+  ## recursion
+  if (depth < 0) {
+    stop("The tree is very large or degenerate.")
+  }
+
+  if (!is.null(parent)) {
+    circ <- check_parents(parent, name)
+    if (circ) {
+      tree$message <- "There appears to be a circular dependency."
+      return(tree)
+    }
+  }
+
+  # make sure a report with this name exists
+  db_reports <- DBI::dbGetQuery(con, "SELECT name FROM report")
+  if (!(name %in% db_reports$name)) {
+    stop("This report does not exist")
```

Comment:
Can this happen? If we're working purely with the DB then the foreign keys are doing all the referential integrity work for us.

Reply:
I think it can only be hit if the user passes in an invalid name. I've moved this logic to *orderly_build_dep_tree*, so we only check once when before the recursive steps begin.